middle_end/flambda2/to_cmm/to_cmm_primitive.ml

```diff
@@ -247,27 +247,65 @@ let array_set_128 ~dbg ~element_width_log2 ~has_custom_ops arr index new_value =
   in
   C.unaligned_set_128 arr index new_value dbg
 
-let array_load ~dbg (kind : P.Array_kind.t)
-    (accessor_width : P.array_accessor_width) ~arr ~index =
+let accessor_assumes_non_custom_block dbg index array_kind =
+  if P.Array_kind.has_custom_ops array_kind
+  then C.add_int_caml index (C.int_const dbg 1) dbg
+  else index
+
+let accessor_assumes_custom_block dbg index array_kind =
+  if P.Array_kind.has_custom_ops array_kind
+  then index
+  else C.sub_int_caml index (C.int_const dbg 1) dbg
+
+let array_load ~dbg (array_kind : P.Array_kind.t)
+    (load_kind : P.Array_load_kind.t) ~arr ~index =
   (* CR mshinwell: refactor this function in the same way as [block_load] *)
-  match kind, accessor_width with
-  | Immediates, Scalar -> C.int_array_ref arr index dbg
-  | (Naked_int64s | Naked_nativeints), Scalar ->
-    C.unboxed_int64_or_nativeint_array_ref arr index dbg
-  | Values, Scalar -> C.addr_array_ref arr index dbg
-  | Naked_floats, Scalar ->
-    C.unboxed_float_array_ref Mutable ~block:arr ~index dbg
-  | Naked_float32s, Scalar -> C.unboxed_float32_array_ref arr index dbg
-  | Naked_int32s, Scalar -> C.unboxed_int32_array_ref arr index dbg
-  | (Immediates | Naked_floats), Vec128 ->
+  match array_kind, load_kind with
+  | ( ( Values | Immediates | Naked_floats | Naked_int32s | Naked_float32s
+      | Naked_int64s | Naked_nativeints | Naked_vec128s ),
+      Immediates ) ->
+    C.int_array_ref arr
+      (accessor_assumes_non_custom_block dbg index array_kind)
+      dbg
+  | ( ( Values | Immediates | Naked_floats | Naked_int32s | Naked_float32s
+      | Naked_int64s | Naked_nativeints | Naked_vec128s ),
+      (Naked_int64s | Naked_nativeints) ) ->
```

Comment:
why isn't it an error to do a read of a `Naked_int64` or `Naked_nativeints` on an array of values (or immediates) ? I'd expect the read to agree with the array_kind (or at least be compatible), and then if need be some conversion to be done (if one really wants to do that kinds of reads in the source code).